### PR TITLE
MCS sreader

### LINKED ISSUE #1167: [TASK] Creation of a MCS Post-Processor
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**

It is needed to import a set of Minimal Cut Sets (MCSs) generated by an external PRA code into RAVEN 

**Describe the solution you'd like**

Read from a csv file the list of MCSs and import them in RAVEN as a PointSet

**Describe alternatives you've considered**

A possible evolution is to import such list of MCSs as a model directly and compute internally the probability associated to all MCSs
----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by mandd**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

Issue #1167 [TASK] Creation of a MCS Post-Processor 
Close #1167 
##### What are the significant changes in functionality due to this change request?

It is needed to import a set of Minimal Cut Sets (MCSs) generated by an external PRA code into RAVEN. It reads from a csv file the list of MCSs and import them in RAVEN as a PointSet.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by wangcj05** [comment]
@mandd The doc test under qsubs failed, could you fix it?

File: framework/PostProcessors/__init__.py [hunk 1]
```diff
@@ -49,6 +49,7 @@
 from .DataClassifier import DataClassifier
 from .ComparisonStatisticsModule import ComparisonStatistics
 from .RealizationAverager import RealizationAverager
+from .MCSimporter import MCSimporter
```
Comment (wangcj05) [inline framework/PostProcessors/__init__.py]: MCSimporter --> MCSImporter?

File: framework/PostProcessors/MCSimporter.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+import csv
+#External Modules End-----------------------------------------------------------
+
+#Internal Modules---------------------------------------------------------------
+from .PostProcessor import PostProcessor
+from utils import InputData
+from utils import xmlUtils as xmlU
+from utils import utils
+import Files
+import Runners
```
Comment (wangcj05) [inline framework/PostProcessors/MCSimporter.py]: Do you need to import Files and Runners?

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml [hunk 3]
```diff
@@ -0,0 +1,58 @@
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/PostProcessors/MCSimporterPostProcessor.MC_importer_Saphire</name>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml]: Two things here:
First, MC_importer_Saphire is used for both test
Second, the "tests" file is not included, in this case, these tests are not tested by the regression test system

File: doc/user_manual/postprocessor.tex [hunk 4]
```diff
@@ -1800,6 +1800,127 @@ \subsubsection{RavenOutput}
 \end{lstlisting}
 The resulting PointSet has \emph{time} as an input and \emph{first} as an output.
 
+
+%%%%%%%%%%%%%% MCSImporter PP %%%%%%%%%%%%%%%%%%%
+
+\subsubsection{MCSimporter}
```
[the remark this answers is not in the document]
Comment (mandd) [reply doc/user_manual/postprocessor.tex]: edited into 'MCS Importer'

File: doc/user_manual/postprocessor.tex [hunk 5]
```diff
@@ -1800,6 +1800,127 @@ \subsubsection{RavenOutput}
...
+\label{MCSimporterPP}
+The \textbf{MCSimporter} post-processor has been designed to import Minimal Cut Sets (MCSs) into RAVEN.
```
[the remark this answers is not in the document]
Comment (mandd) [reply doc/user_manual/postprocessor.tex]: edited into MCImporter

File: doc/user_manual/postprocessor.tex [hunk 6]
```diff
@@ -1800,6 +1800,127 @@ \subsubsection{RavenOutput}
...
+This post-processor reads a csv file which contain the list of MCSs and it save this list as a dataObject
```
[the remark this answers is not in the document]
Comment (mandd) [reply doc/user_manual/postprocessor.tex]: edited

File: doc/user_manual/postprocessor.tex [hunk 7]
```diff
@@ -1800,6 +1800,127 @@ \subsubsection{RavenOutput}
...
+		0 & 0 & 0 & 1 & 0 & 1 & 1.8E-2 & 1 \\
+		0 & 1 & 0 & 0 & 0 & 2 & 4.0E-3 & 1 \\
+		1 & 0 & 1 & 0 & 0 & 3 & 3.0E-4 & 1 \\
+		0 & 0 & 1 & 0 & 1 & 4 & 4.0E-3 & 1 \\
+		\hline
+	\end{tabular}
+\end{table}
+
+%
+\ppType{MCSimporter}{MCSimporter}
```
[the remark this answers is not in the document]
Comment (mandd) [reply doc/user_manual/postprocessor.tex]: edited

File: framework/PostProcessors/MCSimporter.py (hunk 2 above)
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: removed

File: framework/PostProcessors/MCSimporter.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+#Internal Modules End-----------------------------------------------------------
+
+
+class MCSimporter(PostProcessor):
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: fixed

File: framework/PostProcessors/MCSimporter.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+  def _handleInput(self, paramInput):
+    """
+      Method that handles PostProcessor parameter input block.
+      @ In, paramInput, ParameterInput, the already parsed input.
+      @ Out, None
+    """
+    #fileFormat = paramInput.findFirst('fileFormat')
+    #self.fileFormat = fileFormat.value
+    #if self.fileFormat not in self.allowedFormats:
+    #  self.raiseAnError(IOError, 'MCSimporterPostProcessor Post-Processor ' + self.name + ', format ' + str(self.fileFormat) + ' : is not supported')
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: removed

File: framework/PostProcessors/MCSimporter.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+    expand = paramInput.findFirst('expand')
+    self.expand = expand.value
+    # if self.expand = False then the dataObject includes only the Basic Events  listed in the set of MCSs
+    # if self.expand = True then the dataObject includes all Basic Events
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: removed

File: framework/PostProcessors/MCSimporter.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+    if self.expand == True:
+      beListColumn = paramInput.findFirst('BElistColumn')
+      self.BElistColumn = beListColumn.value
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: edited

File: framework/PostProcessors/MCSimporter.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+  def run(self, inputs):
+    """
+      This method executes the PostProcessor action.
+      @ In,  inputs, list, list of file objects
+      @ Out, None
+    """
+
+    MCSfileFound = False
+    BEfileFound  = False
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: edited

File: framework/PostProcessors/MCSimporter.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+    for file in inputs:
+      if file.getType()=="MCSlist":
+        if MCSfileFound:
+          self.raiseAnError(IOError, 'MCSimporterPostProcessor Post-Processor ' + self.name + ', Multiple files with type=MCSlist have been found')
+        else:
+          MCSlistFile = file
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: edited

File: framework/PostProcessors/MCSimporter.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+          MCSfileFound = True
+      if file.getType()=="BElist":
+        if self.expand==False:
+          self.raiseAnError(IOError, 'MCSimporterPostProcessor Post-Processor ' + self.name + ', A file with type=BElist has been found but expand is set to False')
+        if BEfileFound:
+          self.raiseAnError(IOError, 'MCSimporterPostProcessor Post-Processor ' + self.name + ', Multiple files with type=BElist have been found')
+        else:
+          BElistFile = file
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: edited

File: framework/PostProcessors/MCSimporter.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+          BEfileFound  = True
+
+    if BEfileFound==False and self.expand==True:
+      self.raiseAnError(IOError, 'MCSimporterPostProcessor Post-Processor ' + self.name + ', Expand is set to False but no file with type=BElist has been found')
+
+    self.MCSlist=[]
+    self.BElist=set()
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: fixed

File: framework/PostProcessors/MCSimporter.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    self.probability = np.zeros((0))
+    self.MCS_IDs = np.zeros((0))
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: fixed

File: framework/PostProcessors/MCSimporter.py [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+
+        for element in elementsList:
+          element.rstrip('\n')
+        self.MCSlist.append(elementsList)
+        counter = counter+1
+        if self.expand==False:
+          self.BElist.update(elementsList)
+    if self.expand==True:
+      BEdata = pd.read_csv(BElistFile.getFilename())
+      self.BElist = BEdata[self.BElistColumn].values.tolist()
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: fixed

File: framework/PostProcessors/MCSimporter.py [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+        elementsList.pop(0)
+
+        for element in elementsList:
+          element.rstrip('\n')
+        self.MCSlist.append(elementsList)
+        counter = counter+1
+        if self.expand==False:
+          self.BElist.update(elementsList)
+    if self.expand==True:
+      BEdata = pd.read_csv(BElistFile.getFilename())
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: fixed

File: framework/PostProcessors/MCSimporter.py [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+      self.BElist = BEdata[self.BElistColumn].values.tolist()
+
+    MCSpointSet = {}
```
[the remark this answers is not in the document]
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: edited

File: framework/PostProcessors/__init__.py (hunk 1 above)
Comment (mandd) [reply framework/PostProcessors/__init__.py]: edited

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml (hunk 3 above)
Comment (mandd) [reply tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml]: item 1 edited and added tests file

File: doc/user_manual/postprocessor.tex [hunk 20]
```diff
@@ -1800,6 +1800,127 @@ \subsubsection{RavenOutput}
...
+\end{itemize}
+
+\textbf{Example:}
+\begin{lstlisting}[style=XML,morekeywords={anAttribute},caption=MCS Importer input example (no expand)., label=lst:MCS_PP_InputExample]
+  <Files>
+    <Input name="MCSlistFile" type="MCSlist">MCSlist.csv</Input>
+  </Files>
+  
+  <Models>
+    <PostProcessor name="MCSimporter" subType="MCSimporter">
```
[the remark this answers is not in the document]
Comment (mandd) [reply doc/user_manual/postprocessor.tex]: fixed

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml [hunk 21]
```diff
@@ -1,11 +1,11 @@
 <Simulation verbosity="debug">
   <TestInfo>
-    <name>framework/PostProcessors/MCSimporterPostProcessor.MC_importer_Saphire</name>
+    <name>framework/PostProcessors/MCSimporterPostProcessor.MC_importer</name>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml]: MC_importer should be changed to __MCSimporter__. The name should be consistent with the 'tests' file.

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml [hunk 22]
```diff
@@ -1,11 +1,11 @@
 <Simulation verbosity="debug">
   <TestInfo>
-    <name>framework/PostProcessors/MCSimporterPostProcessor.MC_importer_Saphire</name>
+    <name>framework/PostProcessors/MCSimporterPostProcessor.MC_importer_noExpand</name>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml]: MC_importer_noExpand should be changed to __MCSimporterNoExpand__, the name used in __tests__  file.

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml (hunk 21 above)
Comment (mandd) [reply tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml]: Fixed!

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml (hunk 22 above)
Comment (mandd) [reply tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml]: Fixed!

Comment (wangcj05) [reply tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter_noExpand.xml]: It seems to me the change is not consistent with the __tests__ file. __MCSimporterNoExpand__?

**Comment by mandd** [comment]
@alfoa could you take a look at this PR?

File: framework/PostProcessors/MCSimporter.py [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    """
+      Method that handles PostProcessor parameter input block.
+      @ In, paramInput, ParameterInput, the already parsed input.
+      @ Out, None
+    """
+
+    expand = paramInput.findFirst('expand')
+    self.expand = expand.value
+
+    if self.expand == True:
```
Comment (alfoa) [inline framework/PostProcessors/MCSimporter.py]: ```if self.expand:```

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+  <OutStreams>
+    <Print name="PrintPS">
+      <type>csv</type>
+      <source>MCS_PS</source>
+    </Print>
+  </OutStreams>
+  
+  <DataObjects>
+    <PointSet name="MCS_PS">
+      <Input> ACCUMLATOR 1 DISCHARGE CKV 001 FAILS TO OPEN,ACCUMLATOR 2 DISCHARGE CKV 002 FAILS TO OPEN,ACCUMLATOR 3 DISCHARGE CKV 003 FAILS TO OPEN,
```
Comment (alfoa) [inline tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml]: ?

are those the variables?  e.g. ```ACCUMLATOR 1 DISCHARGE CKV 001 FAILS TO OPEN```

**Comment by alfoa** [comment]
1 comment and 1 change

File: tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml (hunk 24 above)
Comment (mandd) [reply tests/framework/PostProcessors/MCSimporterPostProcessor/test_MCSimporter.xml]: yes, these are examples of basic events

File: framework/PostProcessors/MCSimporter.py (hunk 23 above)
Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: edited!

Comment (mandd) [reply framework/PostProcessors/MCSimporter.py]: i also edited the same error in a different line

**Comment by mandd** [comment]
@alfoa i think this PR is good to go

**Comment by alfoa** [comment]
Checklist passed...

PR approved...

**Comment by alfoa** [comment]
Merging..

**Pull request merged**